homeassistant/components/binary_sensor/w800rf32.py

```diff
@@ -0,0 +1,161 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['w800rf32']
+
+ATTR_FIRE_EVENT = 'fire_event'
+EVENT_BUTTON_PRESSED = 'button_pressed'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICES, default={}): {
+        cv.string: vol.Schema({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_FIRE_EVENT, default=False): cv.boolean,
+            vol.Optional(CONF_OFF_DELAY):
+            vol.Any(cv.time_period, cv.positive_timedelta)
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup_platform(hass, config,
+                               add_entities, discovery_info=None):
+    """Set up the Binary Sensor platform to w800rf32."""
+    binary_sensors = []
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary)
+    # --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        _LOGGER.debug("Add %s w800rf32.binary_sensor (class %s)",
+                      entity[ATTR_NAME], entity.get(CONF_DEVICE_CLASS))
+
+        device = W800rf32BinarySensor(
+            hass, device_id, entity.get(CONF_NAME),
+            entity.get(CONF_DEVICE_CLASS),
+            entity[CONF_FIRE_EVENT], entity.get(CONF_OFF_DELAY))
+
+        binary_sensors.append(device)
+
+    add_entities(binary_sensors)
+
+
+class W800rf32BinarySensor(BinarySensorDevice):
+    """A representation of a w800rf32 binary sensor."""
+
+    def __init__(self, hass, device_id, name, device_class=None,
+                 should_fire=False, off_delay=None):
+        """Initialize the w800rf32 sensor."""
+        self._hass = hass
+        self._device_id = device_id.lower()
+        self._signal = W800RF32_DEVICE.format(self._device_id)
+        self._name = name
+        self._should_fire_event = should_fire
+        self._device_class = device_class
+        self._off_delay = off_delay
+        self._state = False
+        self._delay_listener = None
+
+    def _off_delay_listener(self, now):
+        """Switch device off after a delay."""
+        self._delay_listener = None
+        self.update_state(False)
+
+    @property
+    def name(self):
+        """Return the device name."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def should_fire_event(self):
+        """Return is the device must fire event."""
+        return self._should_fire_event
+
+    @property
+    def device_class(self):
+        """Return the sensor class."""
+        return self._device_class
+
+    @property
+    def off_delay(self):
+        """Return the off_delay attribute value."""
+        return self._off_delay
+
+    @property
+    def is_on(self):
+        """Return true if the sensor state is True."""
+        return self._state
+
+    def binary_sensor_update(self, event):
```

Comment:
We can decorate this with `@callback` imported from core.py. This will make it execute in the event loop in async context.